odk/collect/collect_app/src/main/java/org/odk/collect/android/tasks/DownloadFormsTask.java

```diff
@@ -53,34 +58,54 @@ public DownloadFormsTask(FormDownloader formDownloader) {
-
-                formDownloader.downloadForm(serverFormDetails, count -> {
-                    String message = getLocalizedString(Collect.getInstance(), R.string.form_download_progress,
-                            serverFormDetails.getFormName(),
-                            String.valueOf(count),
-                            String.valueOf(serverFormDetails.getManifest().getMediaFiles().size())
-                    );
-
-                    publishProgress(message, currentFormNumber, totalForms);
-                }, this::isCancelled);
-
-                results.put(serverFormDetails, null);
-                FormEventBus.INSTANCE.formDownloaded(serverFormDetails.getFormId());
-            } catch (FormDownloadException.DownloadingInterrupted e) {
-                return emptyMap();
-            } catch (FormDownloadException e) {
-                results.put(serverFormDetails, e);
-                FormEventBus.INSTANCE.formDownloadFailed(serverFormDetails.getFormId(), e.getMessage());
-            }
+            String currentFormNumber = String.valueOf(index);
+            String totalForms = String.valueOf(values[0].size());
+            publishProgress(serverFormDetails.getFormName(), currentFormNumber, totalForms);
+
+            Thread thread = new Thread() {
+                public void run() {
+
+                    try {
+                        formDownloader.downloadForm(serverFormDetails, count -> {
+                            String message = getLocalizedString(Collect.getInstance(), R.string.form_download_progress,
+                                    serverFormDetails.getFormName(),
+                                    String.valueOf(count),
+                                    String.valueOf(serverFormDetails.getManifest().getMediaFiles().size())
+                            );
+
+                            publishProgress(message, currentFormNumber, totalForms);
+
+                        }, DownloadFormsTask.this::isCancelled);
+                    } catch (FormDownloadException e) {
+                        results.put(serverFormDetails, e);
+                        FormEventBus.INSTANCE.formDownloadFailed(serverFormDetails.getFormId(), e.getMessage());
+                        throw new RuntimeException(e);
+                    }
+
+                    new Handler(Looper.getMainLooper()).post(new Runnable() {
+                        @Override
+                        public void run() {
+                            results.put(serverFormDetails, null);
+                            FormEventBus.INSTANCE.formDownloaded(serverFormDetails.getFormId());
+                        }
+                    });
+                }
+            };
+
+            executorService.submit(thread);
 
             index++;
         }
+        executorService.shutdown();
+        // Wait for all tasks to complete
+        try {
+            executorService.awaitTermination(1, TimeUnit.MINUTES);
```

Comment:
On was the basis for choosing 1 minute for timeout? If there are 500 forms, will those be completed in 500 forms?